src/amqproxy/server.cr

```diff
@@ -73,10 +85,41 @@ module AMQProxy
           break
         end
       end
-      @log.info "Proxy stopping accepting connections"
+      wait_for_drain
+    end
+
+    # This function will be called by the close method and at the end of the listen loop, to make sure that
+    # the program exits and the close method only starts closing the socket etc. after all connections are drained.
+    # This means that it is called twice (on SIGTERM/SIGINT and always at the end),
+    # thus we make sure that the logs are only generated on SIGTERM/SIGINT via the `from_close_call` parameter
+    def wait_for_drain(from_close_call = false)
```

Comment:
Why does this method needs to be called twice? (the calls in the in `listen*` methods)

I tried this change and it seems to work? (some changes are from `crystal tool format` that my editor does automatically)

```diff
diff --git a/src/amqproxy/server.cr b/src/amqproxy/server.cr
index 5a12895..b78743c 100644
--- a/src/amqproxy/server.cr
+++ b/src/amqproxy/server.cr
@@ -48,7 +48,7 @@ module AMQProxy
         if client = socket.accept?
           if @draining
             @log.debug "Rejecting new connection #{client.remote_address}"
-            client.close()
+            client.close
             next
           end
           addr = client.remote_address
@@ -57,7 +57,6 @@ module AMQProxy
           break
         end
       end
-      wait_for_drain
     end

     def listen_tls(address, port, cert_path : String, key_path : String)
@@ -70,7 +69,7 @@ module AMQProxy
         if client = socket.accept?
           if @draining
             @log.debug "Rejecting new connection #{client.remote_address}"
-            client.close()
+            client.close
             next
           end
           begin
@@ -85,41 +84,24 @@ module AMQProxy
           break
         end
       end
-      wait_for_drain
     end

-    # This function will be called by the close method and at the end of the listen loop, to make sure that
-    # the program exits and the close method only starts closing the socket etc. after all connections are drained.
-    # This means that it is called twice (on SIGTERM/SIGINT and always at the end),
-    # thus we make sure that the logs are only generated on SIGTERM/SIGINT via the `from_close_call` parameter
-    def wait_for_drain(from_close_call = false)
-      if !@graceful_shutdown
-        if from_close_call
-          @log.info "Skip waiting for open connections to be closed by remote"
-        end
-        return
-      end
-      if from_close_call
-        @log.info "Waiting for open connections to be closed by remote"
-      end
+    def wait_for_drain
+      @draining = true
+      @log.info "Waiting for open connections to be closed by remote"
       while @clients.size > 0
-        if from_close_call
-          @log.info "#{@clients.size} client(s) remaining"
-          @clients.each do |client|
-            @log.debug "- Client #{client.socket.as(TCPSocket).remote_address} since #{(client.lifetime.total_minutes.floor.to_i)}m #{client.lifetime.seconds}s"
-          end
+        @log.info "#{@clients.size} client(s) remaining"
+        @clients.each do |client|
+          @log.debug "- Client #{client.socket.as(TCPSocket).remote_address} since #{(client.lifetime.total_minutes.floor.to_i)}m #{client.lifetime.seconds}s"
         end
-        sleep 20
-      end
-      if from_close_call
-        @log.info "All connections closed. Stopping server"
+        sleep 500.milliseconds
       end
+      @log.info "All connections closed. Stopping server"
     end

     def close
       @log.info "Proxy stopping accepting connections"
-      @draining = true
-      wait_for_drain(true)
+      wait_for_drain if @graceful_shutdown
       @running = false
       @socket.try &.close
       @clients.each &.close
```